### ISSUE TITLE
Repeating Crashes & Performance issues when there is a lot of traffic 

### ISSUE THREAD

**Issue by slowlydev** [slowlydev/f1-dash#57]
We are currently experiencing a lot of issues when there is a on-going session and there is traffic on the website.

Two issues I see are:

- bun crashing with errors I can't fully explain
- using too much bandwidth as f1-dash-data currently sends the full state of timing after every update, which is a design design I regret...

Here are some of the logs regarding the bun crashes:

![image](https://github.com/Slowlydev/f1-dash/assets/61624214/749a75bd-54c6-4910-a35c-4961ca3cd3de)

![image](https://github.com/Slowlydev/f1-dash/assets/61624214/dbe58646-8ac4-46e4-bac9-48028eaee127)

Regarding the bandwidth, I am planning to switch to partial updates, just like the ones we receive from f1.
Which is planned in v2, but that thing is taking a while as I am struggling to find a good clean solution for the backend

**Comment by potts99** [comment]
You could look at using redis?

**Comment by PandelisZ** [comment]
There's sometimes allot of lag between updates as in its not quite real time

If the latency comes down a bit it would make an Amazing trackside companion app as you know exactly what's going on

![20240301_194600.jpg](https://github.com/Slowlydev/f1-dash/assets/8436717/5c97c79f-09a3-4065-b260-8f1e4480e5ce)

Great app and greetings from Bahrain Q3

**Comment by Slowlydev** [comment]
@potts99 Thanks for the comment!

> yup unusable right now, You could look at using redis in order to not overload the requests you are making

Yeah I have noticed, earlier today I added deflate to the bun options of the backend, which relived some of the bandwidth issues, but now as bun / JS is single threaded the "Raspberry Pi" the backend is currently hosted on is struggling at 100% on one core.

And we still get following error:
`error: script "start" was terminated by signal SIGSEGV (Address boundary error) `

Regarding the Redis topic, where and how would you use Redis? Because we only open a WebSocket connection to F1 and then broadcast the messages (messages are currently the full WebSocket state) every time we receive a update from the F1 socket. 

> I would also suggest getting off vercel or at least putting both youre front and backend behind cloudflare. This should eliminate any bandwidth worries

Why, I may ask? bandwidth is not a issue on the frontend side of things, see this screenshot:
(this also includes other projects than f1-dash)   

<img width="313" alt="image" src="https://github.com/Slowlydev/f1-dash/assets/61624214/0f4ae411-1535-4d49-819f-6cd72f92f4f4">

Rather in the backend, the raspberry pi this backend is currently running on only has a Gigabit port and that together with me sending the full state json every time 1,2 or more values change is the issue. 
How would Cloudflare help exactly? Or is there any other "free" hosting for node/bun (in the future, rust) projects?

**Comment by potts99** [comment]
Okay heres what i'd do, the raspberry pi project is fine you just need to elevate the pressure on it. You should be able to get set up with either linode, aws or digital ocean and be able to sign up to a nice free account, digital ocean also have an "app" service that lets you deploy straight to it without the worry for deployment. You should be able to jammy your way to $100/200 dollar accounts with each of these providers. This should let you spin up a much beefier server using the demo credits. 

With regards to vercel, they are great on the hobby tier, but if you go viral, which is probably going to happen with a service like this, the bill can get wild. 

with regards to bun, have you tried using regular node? bun for all its worth in my mind isnt production ready yet.

**Comment by potts99** [comment]
Hows the cpu/ram stats on the pi whilst youre serving these connections? is it even an infra/machine problem? or could this be solved by moving to node?

**Comment by potts99** [comment]
no problems, with keeping it free, you should be able to not have to pay anything for hosting at all as they all offer sign up bonus's

**Comment by vaio2005** [comment]
Jumping here to say that Oracle has a "Forever Free" tier, which gives you 1 to 4 VMs (4 ARM cores and 24GBs of ram, which you can split) and 2 amd-based VMs (but they are slower).

**Comment by Slowlydev** [comment]
@vaio2005 okay thanks Oracle Forever Free looks interesting!

About SSE, yeah might be an idea, how do you think SSE would affect bandwidth? I am asking because I don't know if it has deflate support like WebSockets do.

**Comment by PandelisZ** [comment]
Perhaps this thread is helpful

https://github.com/python-websockets/websockets/issues/968

Particularly the note on compression seem every interesting as you note a large payload size.

Also if you are able to enable github sponsorship I would like to sponsor this project or donate like a server or something :) I'm interesting in infrastructure and scaling of the websockets efficiently is an interesting topic

**Comment by vaio2005** [comment]
> About SSE, yeah might be an idea, how do you think SSE would affect bandwidth? I am asking because I don't know if it has deflate support like WebSockets do.

[I believe they do](https://stackoverflow.com/questions/23769001/is-it-possible-to-use-gzip-compression-with-server-sent-events-sse) and also because they should just be plain HTTP traffic, which does support deflate. 

They lower the amount of overall data as the connection is kept alive, so there is no need to transfer the header like websockets do. 
The tradeoff is slightly higher latency. It may be worth performing some tests/benchmarks to see how they stack up in this case.

**Comment by Slowlydev** [comment]
Okay so @simylein was so nice to rewrite the backend ( https://github.com/Slowlydev/f1-dash/pull/69 ) to use SSE. Let's see how it holds up to this weekend.

**Comment by simylein** [comment]
So now that we had our first run using SSE, we are able to share some performance metrics. 
We had about 200 to 250 concurrent connections for this practice session.
From what I am able to tell we could handle that with ease, even without partial updates. (which of course would enable more headroom)

On the left there is the live log session from the f1-data backend, middle is nmon of our backend and right the reverse proxy with caddy. 
<img width="1728" alt="Screenshot 2024-03-07 at 15 25 18" src="https://github.com/Slowlydev/f1-dash/assets/67859356/babd6084-7ff1-4cd9-8a03-22b5ef7ddd59">

**Comment by vaio2005** [comment]
It was almost 100% smooth on my end during the race! There was only an about 5-6 minute window where the page wasn't loading. Seems like a huge improvement! Do you have any insight on traffic and system load?

**Comment by simylein** [comment]
Yes we do. Here you go. And indeed there is a hard dip there right in the prime time window. 
<img width="1473" alt="Screenshot 2024-03-09 at 20 14 46" src="https://github.com/Slowlydev/f1-dash/assets/67859356/11c146ee-6891-4813-9f3a-96d8706f674b">

**Comment by vaio2005** [comment]
It still looks really good, seems a lot better than the last GP. Did something crash when the dip happened? I guess finding logs isn't difficult since the metrics show pretty clearly when the issue happened.

**Comment by simylein** [comment]
Also if you like reading logs, here is one for you. 
Note logger time is set to UTC, I redacted request and response logs and clean up the we got connections calls for the odd time numbers. For the Best experience view with cat in a terminal because it has colours.
What I figured out so far.
We got disconnected from f1 somehow. 
This started a cascading series of events. 
I failed to properly catch errors for my log flushing functions. 
I got a strong hunch that we somehow managed to start 2 instances because of the many SQLITE_BUSY errors. 
If we didn't it could be that a dead bun process still had the database file open.
[screen.log](https://github.com/Slowlydev/f1-dash/files/14547556/screen.log)

**Comment by vaio2005** [comment]
> I got a strong hunch that we somehow managed to start 2 instances because of the many SQLITE_BUSY errors.    

I was also thinking this, but given the many restarts, I think that  

> it could be that a dead bun process still had the database file open.

is more likely. The backend (which I assume these are the logs of) restarted multiple times, and some of them appears to be manual (SIGKILL - may I ask why this was done? In the first one it appeared to be working fine, serving over 1.2k connections)

Maybe it is worth restarting the websocket after 2-3 seconds of errors?

Also, all the errors appear to happen after 1K connections (apart from the restarts made because of the `SQLITE_BUSY`). I am not an expert of this, but it may be related to that? (I could see the issue happening because if you're using a reverse proxy, all requests to the backend host would come from one IP. If I remember correctly, there is a limit on the amount of open connections at a given time, but that limit is per-IP).  

Lastly, I couldn't find (using GitHub's Code Search) where the message "Was there a typo in the url or port?" originates from. That may help to figure out why the web socket crashes.

**Comment by simylein** [comment]
`Was there a typo in the url or port?` is bun's way of saying you are either offline or the other host could not be reached. Basically a catch all we could not open the connection. 
`Maybe it is worth restarting the websocket after 2-3 seconds of errors?` We actually do this currently, there is a 2 seconds process which checks if there is a connection and if not tries to connect again. (It also checks how many SSE connections there are and only opens the socket if it's more than 0 and closes it if it is 0). 
The rate limiting from fluxify trust the proxy (caddy) which forwards the ip address this should not be a problem. 
I am not sure if caddy has problems over 1k connections but so far I couldn't find any limiters in their docs. 
Lastly, if someone really really likes reading logs, here is the full log session without cleaning from 16:00 to 19:00. (mostly because we can). IP addresses have been obscured and only the first and last two digits are visible. (Also if you view this in a terminal, be aware of the max line buffer, mine is set to 10k lines, so I need to view this bit by bit using head, tail, cat and grep)
[screen.log](https://github.com/Slowlydev/f1-dash/files/14547691/screen.log)

**Comment by simylein** [comment]
And to the `SIGKILL` thingy: We reran our deploy script during the time when we thought it was down. This log should be from that time window. If it's not, we messed up badly and killed a live process serving 1200+ connections.

**Comment by simylein** [comment]
I opened a pr #79 which "fixes" (silences) the `SQLITE_BUSY` errors we got.
